File: typescript/src/tools/assess.ts
import { apiRequest, type RateLimitInfo } from "../client.js";
import { mcpError } from "../errors.js";
import type { RawAssessResponse } from "../schema.js";
import { makeSummary, type AssessSummary } from "../summary.js";
import { assessInputSchema, type AssessInput } from "../types.js";

export const name = "toolrate_assess";

export const config = {
  title: "Assess a tool's reliability before calling it",
  description: [
    "Get ToolRate's reliability score (0-100), pitfalls, mitigations, and the best alternatives for any third-party tool/API the agent is about to call.",
    "",
    "Guard pattern: call this BEFORE invoking the actual third-party tool. After the real call returns, call toolrate_report.",
    "",
    "For LLM provider selection, prefer toolrate_route_llm — it's a simpler interface that returns just the routing fields.",
    "",
    "Quota tip: assess once per tool per workflow, not per call. The API caches 5min server-side. Use toolrate_my_usage to check remaining quota.",
  ].join("\n"),
  inputSchema: assessInputSchema,
};

export async function handler(input: AssessInput) {
  try {
    const { body, rateLimit } = await apiRequest<RawAssessResponse>({
      method: "POST",
      path: "/v1/assess",
      body: {
        tool_identifier: input.tool_identifier,
        context: input.context,
        max_price_per_call: input.max_price_per_call,
        max_monthly_budget: input.max_monthly_budget,
        expected_calls_per_month: input.expected_calls_per_month,
        expected_tokens: input.expected_tokens,
        task_complexity: input.task_complexity,
        budget_strategy: input.budget_strategy,
        eu_only: input.eu_only,
        gdpr_required: input.gdpr_required,
      },
    });

    const trimmed = trimResponse(body, input.verbose === true);
    const summary = makeSummary(body);
    const result: TrimmedAssess = { summary, ...trimmed };
    if (rateLimit) result._meta = { rate_limit: rateLimit };

    return {
      content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }],
    };
  } catch (err) {
    return mcpError(err);
  }
}

interface TrimmedAssess {
  summary: AssessSummary;
  reliability_score: number;
  confidence: number;
  data_source: string;
  historical_success_rate: string;
  predicted_failure_risk: string;
  trend: { direction: string; change_24h: number | null } | null;
  common_pitfalls: Array<{ category: string; percentage: number; mitigation: string | null }>;
  recommended_mitigations: string[];
  top_alternatives: RawAssessResponse["top_alternatives"];
  latency: { p50: number | null; p95: number | null; p99: number | null } | null;
  hosting_jurisdiction: string | null;
  gdpr_compliant: boolean;
  data_residency_risk: string;
  price_per_call: number | null;
  pricing_model: string | null;
  cost_adjusted_score: number | null;
  estimated_monthly_cost: number | null;
  within_budget: boolean | null;
  budget_explanation: string | null;
  recommended_model: string | null;
  reasoning: string | null;
  last_updated: string;
  jurisdiction_source?: string | null;
  jurisdiction_confidence?: string | null;
  jurisdiction_notes?: string | null;
  recommended_for?: string[];
  eu_alternatives?: RawAssessResponse["eu_alternatives"];
  _meta?: { rate_limit: RateLimitInfo };
}

function trimResponse(a: RawAssessResponse, verbose: boolean): Omit<TrimmedAssess, "summary" | "_meta"> {
  const base: Omit<TrimmedAssess, "summary" | "_meta"> = {
    reliability_score: a.reliability_score,
    confidence: a.confidence,
    data_source: a.data_source,
    historical_success_rate: a.historical_success_rate,
    predicted_failure_risk: a.predicted_failure_risk,
    trend: a.trend ? { direction: a.trend.direction, change_24h: a.trend.change_24h } : null,
    common_pitfalls: (a.common_pitfalls ?? []).map((p) => ({
      category: p.category,
      percentage: p.percentage,
      mitigation: p.mitigation,
    })),
    recommended_mitigations: a.recommended_mitigations ?? [],
    top_alternatives: a.top_alternatives ?? [],
    latency: a.latency ? { p50: a.latency.p50, p95: a.latency.p95, p99: a.latency.p99 } : null,
    hosting_jurisdiction: a.hosting_jurisdiction,
    gdpr_compliant: a.gdpr_compliant,
    data_residency_risk: a.data_residency_risk,
    price_per_call: a.price_per_call,
    pricing_model: a.pricing_model,
    cost_adjusted_score: a.cost_adjusted_score,
    estimated_monthly_cost: a.estimated_monthly_cost,
    within_budget: a.within_budget,
    budget_explanation: a.budget_explanation,
    recommended_model: a.recommended_model,
    reasoning: a.reasoning,
    last_updated: a.last_updated,
  };

  if (verbose) {
    base.jurisdiction_source = a.jurisdiction_source;
    base.jurisdiction_confidence = a.jurisdiction_confidence;
    base.jurisdiction_notes = a.jurisdiction_notes;
    base.recommended_for = a.recommended_for ?? [];
    base.eu_alternatives = a.eu_alternatives ?? [];
  }

  return base;
}

File: python/src/toolrate_mcp/tools/assess.py
"""``toolrate_assess`` — crown-jewel reliability scoring tool."""
from __future__ import annotations

import json
from typing import Annotated, Any, Literal, Optional

from pydantic import Field

from ..client import api_request
from ..errors import raise_tool_error
from ..summary import make_summary

NAME = "toolrate_assess"

DESCRIPTION = "\n".join(
    [
        "Get ToolRate's reliability score (0-100), pitfalls, mitigations, and the best alternatives for any third-party tool/API the agent is about to call.",
        "",
        "Guard pattern: call this BEFORE invoking the actual third-party tool. After the real call returns, call toolrate_report.",
        "",
        "For LLM provider selection, prefer toolrate_route_llm — it's a simpler interface that returns just the routing fields.",
        "",
        "Quota tip: assess once per tool per workflow, not per call. The API caches 5min server-side. Use toolrate_my_usage to check remaining quota.",
    ]
)

TaskComplexity = Literal["low", "medium", "high", "very_high"]
BudgetStrategy = Literal["reliability_first", "balanced", "cost_first", "speed_first"]


async def handler(
    tool_identifier: Annotated[
        str,
        Field(
            min_length=1,
            max_length=512,
            description=(
                "URL or canonical name of the tool to assess. Examples: "
                "'https://api.stripe.com/v1/charges', 'sendgrid'."
            ),
        ),
    ],
    context: Annotated[
        Optional[str],
        Field(
            default=None,
            max_length=1024,
            description="Workflow context for context-bucketed scoring, e.g. 'high-value payment processing'.",
        ),
    ] = None,
    max_price_per_call: Annotated[
        Optional[float],
        Field(default=None, ge=0, description="USD cap per call. Tools above are flagged within_budget=false (not filtered)."),
    ] = None,
    max_monthly_budget: Annotated[
        Optional[float],
        Field(default=None, ge=0, description="Max USD spend per month. Combines with expected_calls_per_month for the budget check."),
    ] = None,
    expected_calls_per_month: Annotated[
        Optional[int],
        Field(default=None, ge=0, le=100_000_000, description="Expected call volume. Drives estimated_monthly_cost."),
    ] = None,
    expected_tokens: Annotated[
        Optional[int],
        Field(default=None, ge=0, le=1_000_000, description="Total tokens (input+output) per LLM call. Triggers exact per-token math for LLM providers."),
    ] = None,
    task_complexity: Annotated[
        Optional[TaskComplexity],
        Field(default=None, description="Task complexity hint for LLM model picker. low=Haiku/mini, very_high=Opus/4o."),
    ] = None,
    budget_strategy: Annotated[
        Optional[BudgetStrategy],
        Field(default=None, description="How to weigh reliability vs cost vs latency. Default: reliability_first (80/20)."),
    ] = None,
    eu_only: Annotated[
        Optional[bool],
        Field(default=None, description="Filter alternatives to EU-hosted only."),
    ] = None,
    gdpr_required: Annotated[
        Optional[bool],
        Field(default=None, description="Filter alternatives to GDPR-adequate jurisdictions."),
    ] = None,
    verbose: Annotated[
        Optional[bool],
        Field(default=False, description="Include extra fields (jurisdiction details, eu_alternatives, recommended_for) in the response."),
    ] = False,
) -> str:
    try:
        body, rate_limit = await api_request(
            method="POST",
            path="/v1/assess",
            body={
                "tool_identifier": tool_identifier,
                "context": context,
                "max_price_per_call": max_price_per_call,
                "max_monthly_budget": max_monthly_budget,
                "expected_calls_per_month": expected_calls_per_month,
                "expected_tokens": expected_tokens,
                "task_complexity": task_complexity,
                "budget_strategy": budget_strategy,
                "eu_only": eu_only,
                "gdpr_required": gdpr_required,
            },
        )
    except Exception as e:
        raise_tool_error(e)

    trimmed = _trim_response(body, verbose=bool(verbose))
    result: dict[str, Any] = {"summary": make_summary(body), **trimmed}
    if rate_limit is not None:
        result["_meta"] = {"rate_limit": rate_limit}
    return json.dumps(result, indent=2, default=_default)


def _trim_response(a: dict[str, Any], verbose: bool) -> dict[str, Any]:
    trend = a.get("trend")
    latency = a.get("latency")

    base: dict[str, Any] = {
        "reliability_score": a.get("reliability_score"),
        "confidence": a.get("confidence"),
        "data_source": a.get("data_source"),
        "historical_success_rate": a.get("historical_success_rate"),
        "predicted_failure_risk": a.get("predicted_failure_risk"),
        "trend": (
            {"direction": trend.get("direction"), "change_24h": trend.get("change_24h")}
            if isinstance(trend, dict)
            else None
        ),
        "common_pitfalls": [
            {
                "category": p.get("category"),
                "percentage": p.get("percentage"),
                "mitigation": p.get("mitigation"),
            }
            for p in (a.get("common_pitfalls") or [])
        ],
        "recommended_mitigations": a.get("recommended_mitigations") or [],
        "top_alternatives": a.get("top_alternatives") or [],
        "latency": (
            {"p50": latency.get("p50"), "p95": latency.get("p95"), "p99": latency.get("p99")}
            if isinstance(latency, dict)
            else None
        ),
        "hosting_jurisdiction": a.get("hosting_jurisdiction"),
        "gdpr_compliant": a.get("gdpr_compliant"),
        "data_residency_risk": a.get("data_residency_risk"),
        "price_per_call": a.get("price_per_call"),
        "pricing_model": a.get("pricing_model"),
        "cost_adjusted_score": a.get("cost_adjusted_score"),
        "estimated_monthly_cost": a.get("estimated_monthly_cost"),
        "within_budget": a.get("within_budget"),
        "budget_explanation": a.get("budget_explanation"),
        "recommended_model": a.get("recommended_model"),
        "reasoning": a.get("reasoning"),
        "last_updated": a.get("last_updated"),
    }

    if verbose:
        base["jurisdiction_source"] = a.get("jurisdiction_source")
        base["jurisdiction_confidence"] = a.get("jurisdiction_confidence")
        base["jurisdiction_notes"] = a.get("jurisdiction_notes")
        base["recommended_for"] = a.get("recommended_for") or []
        base["eu_alternatives"] = a.get("eu_alternatives") or []

    return base


def _default(obj: Any) -> Any:
    if isinstance(obj, dict):
        return dict(obj)
    raise TypeError(f"Object of type {type(obj).__name__} is not JSON serialisable")
